include/watchdog.h

```diff
@@ -55,62 +71,234 @@ struct wdt_config {
+	void (*get_config)(struct device *dev, struct wdt_config *config);
+	int (*set_config)(struct device *dev, struct wdt_config *config);
+	void (*reload)(struct device *dev);
 };
+/**
+ * @endcond
+ */
 
-static inline void wdt_enable(struct device *dev)
+/**
+ * @brief Set up watchdog instance.
+ *
+ * This function is used for configuring global watchdog settings that
+ * affect all timeouts. It should be called after installing timeouts.
+ * After successful return, all installed timeouts are valid and must be
+ * serviced periodically by callig wdt_feed().
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param options Configuration options as defined by the WDT_OPT_* constants
+ *
+ * @retval 0 If successful.
+ * @retval -ENOTSUP If any of the set options is not supported.
+ * @retval -EBUSY If watchdog instance has been already setup.
+ */
+static inline int wdt_setup(struct device *dev, u8_t options)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->enable(dev);
+	return api->setup(dev, options);
 }
 
-static inline void wdt_disable(struct device *dev)
+/**
+ * @brief Disable watchdog instance.
+ *
+ * This function disables the watchdog instance and automatically uninstalls all
+ * timeouts. To set up a new watchdog, install timeouts and call wdt_setup()
+ * again. Not all watchdogs can be restarted after they are disabled.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 If successful.
+ * @retval -EFAULT If watchdog instance is not enabled.
+ * @retval -EPERM If watchdog can not be disabled directly by application code.
+ */
+static inline int wdt_disable(struct device *dev)
 {
 	const struct wdt_driver_api *api = dev->driver_api;
 
-	api->disable(dev);
+	return api->disable(dev);
 }
 
-static inline void wdt_get_config(struct device *dev,
+/**
+ * @brief Install new timeout.
+ *
+ * This function must be used before wdt_setup(). Changes applied here
+ * have no effects until wdt_setup() is called.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param cfg Pointer to timeout configuration structure.
+ *
+ * @retval 0 If successful.
+ * @retval -EBUSY If timeout can not be installed while watchdog has already
+ *		  been setup.
+ * @retval -ENOMEM If no more timeouts can be installed.
+ * @retval -EFAULT If timeout with given handle has already been installed.
+ *		   Multistage watchdogs are allowed to use the same handle
+ *		   for different stages
+ * @retval -ENOTSUP If any of the set flags is not supported.
+ * @retval -EINVAL If any of the window timeout value is out of possible range.
+ *		   This value is also returned if watchdog supports only one
+ *		   timeout value for all timeouts and the supplied timeout
+ *		   window differs from windows for alarms installed so far.
+ */
+static inline int wdt_install_timeout(struct device *dev,
+				      struct wdt_timeout_cfg *cfg)
```

Comment:
Is there a reason why the second parameter is not a pointer to a constant structure? It seems that this function is not supposed to modify anything in the pointed content. And currently one cannot use a constant structure (e.g. stored in flash) without doing ugly typecasting.

Reply:
Good point. It certainly makes sense to make it const. We've overlooked it.